### PR TITLE
Generalize supervisor

### PR DESCRIPTION
**Pull request by regiskuckaertz**
Solves #124

### CODE REVIEW

**Comment by jdegoes** [comment]
@regiskuckaertz This looks really good. I'd only make the following changes:

1. Make the supervisor mandatory on the `Supervise` constructor to simplify logic inside RTS.
2. Implement `Fiber.interruptAll` which interrupts an iterable of heterogeneous fibers.
3. Implement `Fiber.joinAll` which joins an iterable of heterogeneous fibers (attemping them, so they cannot error, and finally returning `Unit`).
4. Call the `supervise` combinator that takes a supervisor `IO.superviseWith(io)(supervisor)`.
5. In the ordinary `supervise` combinator (which uses a default supervisor), call `IO.superviseWith(io)(Fiber.interruptAll)`.

This will make for a smaller number of more reusable and simpler pieces.

**Comment by regiskuckaertz** [comment]
@jdegoes Done, that's much better indeed 👍

File: core/shared/src/main/scala/scalaz/zio/Fiber.scala [hunk 1]
```diff
@@ -82,4 +82,10 @@ object Fiber {
       def join: IO[E, A]                                     = IO.point(a)
       def interrupt0(ts: List[Throwable]): IO[Nothing, Unit] = IO.unit
     }
+
+  final def interruptAll(fs: Iterable[Fiber[_, _]]): IO[Nothing, Unit] =
+    fs.foldLeft(IO.unit)((io, f) => io *> f.interrupt)
+
+  final def joinAll(fs: Iterable[Fiber[_, _]]): IO[Nothing, Unit] =
+    fs.foldLeft(IO.unit)((io, f) => io *> f.join.attempt.const(()))
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/Fiber.scala]: The code `.const(())` can be replaced by `.toUnit`.

**Comment by jdegoes** [comment]
@regiskuckaertz Looks beautiful! One minor nit (`.toUnit` instead of `.const(())`), and then good to merge!

**Comment by regiskuckaertz** [comment]
Fixes #124

**Pull request merged**